lib/src/source/git.dart

```diff
@@ -584,24 +587,26 @@ class BoundGitSource extends CachedSource {
   ///
   /// If [shallow] is true, creates a shallow clone that contains no history
   /// for the repository.
-  Future _clone(
-    String from,
-    String to, {
-    bool mirror = false,
-    bool shallow = false,
-  }) {
+  ///
+  /// If [reference] is not null, we expect [from] to be a remote repository
+  ///  and [reference] to be the URI of a local filesystem repository.
+  /// This allows our clone to save network and filesystem resources while
+  /// still allowing Git LFS to fetch any binary files from the remote.
+  Future _clone(String from, String to,
+      {bool mirror = false, bool shallow = false, String reference}) {
     return Future.sync(() {
       // Git on Windows does not seem to automatically create the destination
       // directory.
       ensureDir(to);
+      
       var args = [
         'clone',
+        if (reference != null) ['--reference', reference];
```

Comment:
Isn't this the only thing that is necessary?

--------------------------

If `git-lfs` is globally installed with `git lfs install`, then `git clone` just works.. but `git clone path/to/canonical/cache path/to/revision/cache` doesn't work... but if instead we do:

`git clone --reference path/to/canonical/cache https://original-repository-url.com/repo path/to/revision/cache` then git-lfs just works.

It's perhaps not the fastest thing, but if we do it this way we don't need any special handling for `git lfs` in `pub` (it just works). And if we don't need special handling for `git lfs`, then it's more reasonable to argue why we don't need test cases for `git lfs`.

-----------------

Maybe, we can make a warning if we see an error message indicating a `git lfs` repository, but `git lfs` isn't installed globally.


----------------------------------


It's also possible that it's much better to build special logic to install git lfs in the specific repository and do `git lfs pull`. I'm just saying I think the minimal changes might be easier to land.

Reply:
@jerel thoughts? ^

Reply:
@jonasfj if I remember right the current approach of running a `git lfs pull` was the most dependable at fetching updates and/or version changes after the initial fetch attempt. But it would be worth doing some manual testing to see if modifying the regular `git clone` command to one that also supports LFS would work in all cases as that could be an easier implementation and lower maintenance going forward.

Test case 1: what happens if the pubspec is set to:
```
git:
  url: ../some-lib
  ref: feature-branch
```
where `feature-branch` contains LFS tracked blob files. The `some-lib` developer pushes an update to `feature-branch` and the consuming application wants to update to the latest commit. Does `pub get` successfully pull the latest commit and binary files?

Test case 2: using the pubspec above and a large binary file, break the network after the repo has been fetched but while the file is being downloaded. Does the binary get [re]fetched next time `pub get` is ran or does `git clone` `exit 0` without retrying the binary download?

<hr>

If the implementation needs a local LFS server there is one that's part of the LFS project (I haven't used it) https://github.com/git-lfs/lfs-test-server but it doesn't look like there's recent binaries built so it would probably involve building and uploading artifacts managed by this project.